XR Scaffolding + Updates for various effects

robbiesri opened the pull request:
This PR is composed of general XR infrastructure, along with plumbing for specific effects.

**Temporary Texture Allocation through PostProcessRenderContext**
The biggest change here is that temporary screen space target allocation gets routed through the PostProcessRenderContext in order to handle various texture format setups in XR.  XR single-pass can create double-wide textures or texture arrays, depending on the underlying mode, and it makes sense that the PostProcessRenderContext handle this logic.  I think it also 'looks' a bit better to allow the context to manage the particulars of screen space temps.

**Logical Screen Width/Height**
PostProcessRenderContext. provides screenWidth and screenHeight, which correspond to the logical screen size.  This is used for aspect ratios, number of steps in a texture pyramid, screen size in a shader.  This is mostly useful for double-wide textures in single-pass, since the actual texture is for two eyes, but draws still happen for one eye.

**RenderViewportScale Supported**
PostProcessLayer manages the script side portion of RenderViewportScale, and TransformStereoScreenSpaceTex in xrLib.hlsl takes care of the shader logic.

**Explicit Effect Fixups**
- SAO - 
- AutoExposure
- Bloom
- Color Grading
- Depth of Field
- Grain
- TAA

**Known Untested/Unsupported**
- MSVO - Seems to be under active dev, so I laid off
- Motion Blur - Not really XR appropriate (but I could support it)
- SSR - Haven't started port process yet
- SMAA

**Testing**
* Tested all effects with monoscopic, multi-pass, single-pass double-wide on PC.  Tested with a couple test scenes (Shader Calibration, Corridor).
* Not yet tested with Single Pass Instancing, as more work needs to be done.
* Tested on 2017.3.  I need to test the other versions.  Single-pass will only work on 2017.2+, which is already specified in isSinglePassStereoEnabled.

robbiesri committed:
PostProcessRenderContext manages RenderTextureDescriptor

The RenderTextureDescriptor is either sourced from Camera parameters or from XRSettings.eyeTextureDesc.  It also returns a new RenderTextureDescriptor for temporary allocation usage.

robbiesri committed:
Begin usage of RenderTextureDescriptor

PostProcessLayer starts using the RenderTextureDescriptor from PostProcessRenderContext

robbiesri committed:
Move AO over to RenderTextureDesc

AO continues to work.
Single-pass DW has an issue with Lowest to Medium due to the quarter res intermediate buffer.  There is a bug that needs to be fixed in the engine; has to do with viewports.

robbiesri committed:
First pass at AutoExposure XR update

This properly supports multi-pass XR.
For single-pass double-wide, there is the question of a unified histogram buffer across both eyes.  It's not entirely clear to me that we really need individual histogram buffers per eye in single-pass.

Additionally, the LogHistogram workflow has to be updated to at least support Tex2DArray.

Also, I just realized we might need to add Tex2DMS support.  I wonder at what point in the pipeline we want to resolve the texture, and hand it off to post-processing to operate on.

robbiesri committed:
Add XR support to Bloom plumbing

Make sure pyramid generation uses correct texture sizes and # of steps
Correct the screen ratio used to compare against the dirt texture

robbiesri committed:
Support LDR Color Grading in Single-Pass Stereo

Bit of a workaround for now.  A LUT is generated by a draw, which ends up stereo-ized in single-pass.  The problem with this is the LUT isn't a stereo texture, it's a LUT.  The workaround is that each stereo eye generates half of the LUT.

robbiesri committed:
Update DepthOfField for XR

Decent amount of changes:
* Move over to using RenderTextureDescriptor for RT allocs.
* Fix typo bug in FragPrefilter
* Allocate CoCHistoryTextures per eye for Multi-Pass
* Fix aspect ratio for stereo case

robbiesri committed:
Update Grain for XR

Same issue as Color Grading LUT for LDR.  We are creating a non-stereo render texture, but it goes through the render texture path.  So we work around it by baking half the texture per eye.

robbiesri committed:
Update Dithering for XR

Just update the Dithering_Coords with the right screen space args

robbiesri committed:
Merge remote-tracking branch 'origin/xr_taa' into xr_update

Fixed some merge conflicts
Need to amend AO code

robbiesri committed:
Fix bad merge in AO base code

robbiesri committed:
Support ScalableAO in XR

This was originally supported before AO refactor.  Scalable must be less useful one, but it was easy to port back.

robbiesri committed:
Merge branch 'v2' into xr_update

robbiesri committed:
Post-merge cleanups

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_update

robbiesri committed:
Support RenderViewportScale

Traditional screen-space RenderViewportScale support is done through _MainTex_ST.  However, we don't use Graphics.Blit or CommandBuffer.Blit for most blits, so we can't rely on that infrastructure.

Instead, we can wrap our usage of unity_StereoScaleOffset/TransformStereoScreenSpaceTex with an RVS scaler.  This works pretty well, and makes the RenderViewportScale management very explicit and obvious.

There might be an issue with WindowsMR, as it doesn't guarantee the RenderViewportScale value will actually be used (yeah...), but we'll cross that bridge when we get there.

robbiesri committed:
Use texcoordStereo instead of UnityStereoTransformScreenSpaceTex

Couple places I had to port.  Incidentally, in the bakers, boo!

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_update

Lut2DBaker did need manual merge, since it uses i.texcoordStereo

robbiesri committed:
New PostProcessRenderContext.GetScreenSpaceTemporaryRT

We need XRSettings.eyeTextureDesc to create proper single-pass temporary RTs.  But they aren't available until 2017.2.  So if there is the need to create a temporary screen space RT, there is a new function that handles the allocation.

It either uses the old explicit-argument-driven GetTemporaryRT for pre-2017.2 (monoscopic + multi-pass) or RenderTextureDescriptor 2017.2+.  There is also a 'scaler' that is applied to the texture dimensions (if needed).  I might make this more explicit in the future with explicit arguments changing the size.

robbiesri committed:
Partial port to using GetScreenSpaceTemporaryRT

Still parts to do
Also need to create RenderTexture version.  Currently, only command buffer

robbiesri committed:
Create RenderTexture version of GetScreenSpaceTemporaryRT

In the places that create a RenderTexture, they need a non-desc dependent function as well.

Plus, use it in the ScalableAO infrastructure.

robbiesri committed:
Finish port to GetScreenSpaceTemporaryRT

The remaining clients that were getting temporary RTs now use GetScreenSpaceTemporaryRT.  

Also, instead of a dimension scaler, there are explicit overrides for width and height.

robbiesri committed:
Clean up TAA todo

robbiesri committed:
Create and use screen w/h props in PostProcessRenderContext

There is a difference between the logical w/h of a screen and the texture size being allocated.  So I started to formalize the difference with the addition of the screen size tracking.

robbiesri committed:
Quick cleanups

robbiesri committed:
TAA port to GetScreenSpaceTemporaryRT

Kink3d commented:
Good work Rob. Ill test this over the next few days.

robbiesri committed:
Cleanup ScalableAO constants

I accidentally removed some helpful constants in the ScalableAO code.  I'm adding them back in.

robbiesri committed:
Stop querying isDeviceActive, start using Camera.stereoEnabled

XRSettings.isDeviceActive makes sense for general XR activity, but we should be using Camera.stereoEnabled for per camera stereo state.  I do cache this value in the PostProcessRenderContext.

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_update

robbiesri committed:
Take RVS into account for clamp

Clamp needs to know about RVS so it can clamp neighborhoods correctly.

New TODO:  we need to make sure the TAA history textures have previous RVS values accounted for...

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_update

Kink3d commented on PostProcessing/Runtime/PostProcessRenderContext.cs:
```diff
@@ -90,14 +104,93 @@ public Camera camera
+            return RenderTexture.GetTemporary(desc);
+#else
+            int actualWidth = width;
+            int actualHeight = height 
```
Missing semicolon :D

Kink3d commented on PostProcessing/Runtime/PostProcessRenderContext.cs:
```diff
@@ -23,28 +23,42 @@ public Camera camera
+                    if (m_Camera.stereoActiveEye == Camera.MonoOrStereoscopicEye.Right)
                         xrActiveEye = (int)Camera.StereoscopicEye.Right;
 
-                    xrSingleEyeWidth = XRSettings.eyeTextureWidth;
```
RuntimeUtilities.cs L449 needs to reflect this change

Kink3d commented on PostProcessing/Runtime/PostProcessRenderContext.cs:
```diff
@@ -90,14 +104,93 @@ public Camera camera
+            return RenderTexture.GetTemporary(desc);
+#else
+            int actualWidth = width;
+            int actualHeight = height 
```
Missing semicolon :D 
`int actualHeight = height;`

robbiesri replied:
Someone else pointed this to me as well!  How the heck did this ever work?  Thanks, will fix!

Kink3d commented on PostProcessing/Runtime/PostProcessRenderContext.cs:
```diff
@@ -23,28 +23,42 @@ public Camera camera
+                    if (m_Camera.stereoActiveEye == Camera.MonoOrStereoscopicEye.Right)
                         xrActiveEye = (int)Camera.StereoscopicEye.Right;
 
-                    xrSingleEyeWidth = XRSettings.eyeTextureWidth;
```
RuntimeUtilities.cs L449 needs to be updated for this

robbiesri replied:
Nice, thanks.

Kink3d reviewed (changes requested):
Generally good. Some issues I found:
- 5.6 SPS Oculus: there is a tiny duplicate render target in the bottom left of the view (only repros in headset)
- 17.3 SI Oculus - Left eye is single pixel (flat color) but you said this isnt implemented yet
- 17.3 SPS Oculus - Still displays warning for version when TAA enabled
- MSVO is horrible in VR (so jittery and wobbly). This is just how the effect is though, Keijiro is looking at it.
- Plus code comments

robbiesri committed:
PR feedback changes from Matt Dean

robbiesri committed:
Fix TAA + Single Pass warning in Post Process Layer Editor

The warning should only pop up in versions before 2017.3.  2017.3+ is fine, but it was still showing the warning.

robbiesri commented:
OK @Kink3d, I got some updates!

- Fixed the code bits you requested, thanks for catching those.
- Fixed the TAA + SPS warning in the Post Process Layer Editor.  
- MSVO was originally out of the scope of this PR since I saw it was under active development.  Happy to see Keijiro looking at it 🔨 
- Indeed, Stereo Instancing is not implemented yet, so artifacts are expected.  Once this change lands, I can get to work on that 👍 
- 5.6 SPS is not technically supported with this release. I snuck in a line in the description about official SPS support restricted to 2017.2+.  The reason is that there isn't any programmatic way to detect SPS being enabled in the Player, which means we can't create SPS-aware RenderTextures.  I've talked to Thomas about possibly having a manual override for earlier versions where the developer will flip a switch in the PostProcessLayer itself indicating SPS, but that's a future PR (and would require further discussion).  I'm surprised 5.6 only had that artifact!

Let me know if there's anything else to check.  I pushed a couple new changes that should address your feedback!  Thanks again for looking at this!

robbiesri commented:
Going back to the SPS + 5.6.  I assume you were running this from the Editor?  When I look at the code, the Editor version of isSinglePassStereoEnabled just checks for Editor settings, which is fine.

I think I need to add a warning to the PostProcessLayer Editor that notifies the developer if they are trying to use the Stack with 2017.1 or older.  Let me investigate that.  Might just be able to do it with the UNITY_* version defines.

robbiesri committed:
Warning in PostProcessLayer InspectorGUI for SPS and pre-2017.2

SPS is only officially supported in 2017.2+ because of XRSettings.eyeTextureDesc.  There are manual workarounds, but they aren't configured yet.  Might be in the future.

robbiesri commented:
I added a warning to the PostProcessLayer InspectorGUI if pre-2017.2 and SPS are paired up.  That will work until we figure out some sort of compatibility implementation for older Unity versions.

Kink3d commented:
@robbiesri The bool `isSinglePassStereoEnabled` (thats being used for both the SPS & <17.2 and TAA & SPS & <17.3 error messages) only returns when `Application.isPlaying`

Obviously this means neither error message is displayed outside of play mode. Needs to be fixed one way or another (I removed that `isPlaying` check and everything seems fine but idk where else its being used).

robbiesri commented:
Actually, @chman wrote the editor version of isSinglePassStereoEnabled, and put in the original check for TAA.  I stuck with Thomas' version on purpose because I thought he only wanted the message on during Play mode.  And if you think about it, single-pass stereo is really only enabled during Play mode...

Since it's only for the Inspector window warnings, I don't see the harm in adding a version of the check that is independent of Application.isPlaying.  The current version has to stay for all the other uses (so effects don't get disabled in Edit mode).

robbiesri committed:
Add Editor-only construct to check if SPS selected

Used for Inspector window warnings

robbiesri commented:
Ok @Kink3d, I added an Editor-only check that looks for the SPS option being selected, and I use it for the Inspector window generation.  So now you will see warnings in Edit mode, but it won't disable any Edit-mode functionality.

Chman commented on PostProcessing/Runtime/Effects/DepthOfField.cs:
```diff
@@ -84,19 +95,20 @@ float CalculateMaxCoCRadius(int screenHeight)
 
-                rt = RenderTexture.GetTemporary(width, height, 0, format);
-                rt.name = "CoC History";
+                // TODO: The CoCCalculation CoCTex uses RenderTextureReadWrite.Linear, why isn't this?
```
Good catch, will put a note to look into it.

Chman commented on PostProcessing/Runtime/PostProcessLayer.cs:
```diff
@@ -272,8 +272,15 @@ void OnPreCull()
+            if (m_Camera.stereoEnabled)
+            {
                 m_Camera.ResetStereoProjectionMatrices();
+                Shader.SetGlobalFloat("rvsGlobal", XRSettings.renderViewportScale);
```
Please use `Shader.SetGlobalFloat(int, float)` to avoid hashing on every frame. See the `ShaderIDs` file.

robbiesri replied:
Ah, thanks for this pointer.  I'll fix this up!

Chman commented on PostProcessing/Runtime/Utils/RuntimeUtilities.cs:
```diff
@@ -205,6 +205,7 @@ public static void BlitFullscreenTriangle(Texture source, RenderTexture destinat
         // Assumes that both textures have the exact same type and format
         public static void CopyTexture(CommandBuffer cmd, RenderTargetIdentifier source, RenderTargetIdentifier destination)
         {
+            // TODO: Not sure if this works with texture arrays...
```
According to the doc it should.

robbiesri replied:
Yep, I should delete this.  A quick investigation into the underlying graphics APIs seems to confirm this.

Chman commented on PostProcessing/Shaders/xRLib.hlsl:
```diff
@@ -24,6 +24,8 @@ CBUFFER_START(UnityStereoEyeIndex)
 CBUFFER_END
 #endif
 
+float rvsGlobal;
```
Can we name this uniform more explicitely? `rsvGlobal` is a bit vague...

robbiesri replied:
Yeah, I don't know why I used such a crappy name.  I'll use _RenderViewportScaleFactor

Chman reviewed (changes requested):
I requested two minor changes but it looks good overall.

robbiesri committed:
Integrate feedback from @chman

Better name for RenderViewportScale value
Use ShaderIDs instead of raw string (hashing)

robbiesri commented:
Fixed the stuff you asked for @Chman. Thanks for reviewing!

Chman commented:
Awesome PR, thanks!
